node/src/types/value_or_chunk.rs

```diff
@@ -0,0 +1,207 @@
+use std::{
+    fmt::{self, Display, Formatter},
+    marker::PhantomData,
+};
+
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use casper_hashing::{ChunkWithProof, Digest, MerkleConstructionError};
+use casper_types::bytesrepr::{self, Bytes, FromBytes, ToBytes, U8_SERIALIZED_LENGTH};
+use datasize::DataSize;
+
+/// Represents a value or a chunk of data with attached proof.
+/// Chunk with attached proof is used when the requested
+/// value is larger than [ChunkWithProof::CHUNK_SIZE_BYTES].
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub enum ValueOrChunk<V> {
+    /// Represents a value.
+    Value {
+        /// Serialized `V` type using `ToBytes` impl.
+        data: Bytes,
+        /// Phantom marker.
+        _phantom: PhantomData<V>,
+    },
+    /// Represents a chunk of data with attached proof.
+    ChunkWithProof(ChunkWithProof),
+}
+
+#[derive(Debug, PartialEq, Error, Eq)]
+/// Error returned when constructing an instance of `ValueOrChunk`.
+pub enum ChunkingError {
+    /// Merkle proof construction error
+    MerkleConstruction(MerkleConstructionError),
+    /// Serialization error
+    SerializationError(bytesrepr::Error),
+}
+
+impl Display for ChunkingError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            ChunkingError::MerkleConstruction(error) => f
+                .debug_tuple("MerkleConstructionError")
+                .field(&error)
+                .finish(),
+            ChunkingError::SerializationError(error) => {
+                f.debug_tuple("SerializationError").field(&error).finish()
+            }
+        }
+    }
+}
+
+impl<V> ValueOrChunk<V> {
+    /// Creates an instance of `ValueOrChunk::Value` if data is less than `max_chunk_size`
+    /// or a `ValueOrChunk::ChunkWithProof` if it is. In the latter case it will return only the
+    /// `chunk_index`-th chunk of the value's byte representation.
+    pub fn new(data: V, chunk_index: u64, max_chunk_size: usize) -> Result<Self, ChunkingError>
+    where
+        V: ToBytes,
+    {
+        let bytes = ToBytes::into_bytes(data).map_err(ChunkingError::SerializationError)?;
+        if bytes.len() <= max_chunk_size {
+            Ok(ValueOrChunk::Value {
+                data: bytes.into(),
+                _phantom: PhantomData,
+            })
```

Comment:
Should we print a warning and return an error here if `chunk_index != 0`?  Seems like that would be the correct thing to do in case we were handling a malformed/malicious request maybe?

Reply:
We will return it inside the [`ChinkWithProof::new`](https://github.com/casper-network/casper-node/blob/dev/hashing/src/chunk_with_proof.rs#L79)